flang/lib/Optimizer/Transforms/ArrayValueCopy.cpp

```diff
@@ -426,6 +426,93 @@ static bool hasPointerType(mlir::Type type) {
   return type.isa<fir::PointerType>();
 }
 
+// This is a NF performance hack. It makes a simple test that the slices of the
+// load, \p ld, and the merge store, \p st, are trivially mutually exclusive.
+static bool mutuallyExclusiveSliceRange(ArrayLoadOp ld, ArrayMergeStoreOp st) {
+  // If the same array_load, then no further testing is warranted.
+  if (ld.getResult() == st.original())
+    return false;
+
+  auto getSliceOp = [](mlir::Value val) -> fir::SliceOp {
+    if (!val)
+      return {};
+    auto sliceOp = mlir::dyn_cast_or_null<fir::SliceOp>(val.getDefiningOp());
+    if (!sliceOp)
+      return {};
+    return sliceOp;
+  };
+
+  auto ldSlice = getSliceOp(ld.slice());
+  auto stSlice = getSliceOp(st.slice());
+  if (!ldSlice || !stSlice)
+    return false;
+
+  // Resign on subobject slices.
+  if (!ldSlice.fields().empty() || !stSlice.fields().empty() ||
+      !ldSlice.substr().empty() || !stSlice.substr().empty())
+    return false;
+
+  // Crudely test that the two slices do not overlap by looking for the
+  // following general condition. If the slices look like (i:j) and (j+1:k) then
+  // these ranges do not overlap. The addend must be a constant.
+  auto ldTrips = ldSlice.triples();
+  auto stTrips = stSlice.triples();
+  const auto size = ldTrips.size();
+  if (size != stTrips.size())
+    return false;
+
+  auto displacedByConstant = [](mlir::Value v1, mlir::Value v2) {
+    auto removeConvert = [](mlir::Value v) -> mlir::Operation * {
+      auto *op = v.getDefiningOp();
+      while (auto conv = mlir::dyn_cast_or_null<fir::ConvertOp>(op))
+        op = conv.value().getDefiningOp();
```

Comment:
Is it possible that one of the iterations of this loop sets `op` to `true` and a subsequent one sets it to `false`?

Reply:
The loop removed intervening `fir.convert` operations of type `mlir::Operation *`. So if we have a chain of something like

```mlir
%1 = ...
%2 = fir.convert %1 ...
%3 = fir.convert %2 ...
%4 = fir.convert %3 ...
```

The calling this with `%4` returns `%1`. Note that this is done on array index expressions where we want to "skip over" all the conversions to machine word integers, etc.